board.py

```diff
@@ -2,79 +2,170 @@
+        """Parse food from food_data."""
+        return set([Cell(*food) for food in food_data])
+
+    def parse_snakes(self, snakes_data):
+        """Parse snakes from snakes_data."""
+        return {snake_data["id"]: Snake(snake_data)
+                for snake_data in snakes_data}
+
+    def parse_dead_snakes(self, dead_snakes_data):
+        """Parse dead snakes from dead_snakes_data."""
+        return {dead_snake_data["id"]: Snake(dead_snake_data)
+                for dead_snake_data in dead_snakes_data}
+
+    def build_grid(self):
+        """Build a grid of EntityId that represents the current board state."""
+        self.board = [[EntityId.EMPTY for _ in range(self.width)] for _ in range(self.height)]
+
+        # Mark food
+        for cell in self.food:
+            self._set_entity_at_cell(cell, EntityId.FOOD)
+
+        # Mark live snakes
+        for cell in self.snakes.body:
+            self._set_entity_at_cell(cell, EntityId.BODY)
+        self._set_entity_at_cell(self.snakes.head, EntityId.HEAD)
+
+    def is_snake_id_at_cell(self, snake_id, cell):
+        """Return True if Entity at Cell is a Snake and has snake_id."""
+        for snake in self.snakes:
+            if snake.id is snake_id:
+                return cell in snake.head + snake.body
+        return False
+
+    def get_entity_at_cell(self, cell):
+        """Get EntityId at Cell."""
+        return self.get_entity_at_pos(cell.x_pos, cell.y_pos)
+
+    def get_entity_at_pos(self, x_pos, y_pos):
+        """Get EntityId at position."""
+        return self.board[y_pos - 1][x_pos]
+
+    def _set_entity_at_cell(self, cell, entity_id):
+        """Set EntityId at Cell."""
+        self._set_entity_at_pos(cell.x_pos, cell.y_pos, entity_id)
+
+    def _set_entity_at_pos(self, x_pos, y_pos, entity_id):
+        """Set EntityId at position."""
+        self.board[y_pos - 1][x_pos - 1] = entity_id
+
+    def cell_within_bounds(self, cell):
+        """Return True if Cell is within bounds of the game grid."""
+        return self.pos_within_bounds(cell.x_pos, cell.y_pos)
+
+    def pos_within_bounds(self, x_pos, y_pos):
+        """Return True if position is within bounds of the game grid."""
+        return 0 <= x_pos - 1 <= self.width and 0 <= y_pos - 1 <= self.height
+
+    def update_board(self, snake_id_move_mapping):
+        """State transition for board. Applies moves to all live snakes.
+        * Grid is invalidated until the end of this method *
+        1. Move snakes
+        2. Snakes dying
+        3. Update world
+        4. Update grid"""
+
+        self.move_snakes(snake_id_move_mapping)
+        self.resolve_deaths()
+        self.update_board_after_move()
+        self.build_grid()
+
+    def move_snakes(self, snake_id_move_mapping):
+        """Apply moves to snakes."""
+        for snake_id, move in snake_id_move_mapping.item():
+            self.snakes[snake_id].apply_move_to_snake(move)
+
+    def resolve_deaths(self):
```

Comment:
This logic is kind of unclear here. Can the `_dead` and `_live` have better names?